modules/llm-cache/ds/kv_cache_chunk.h

```diff
@@ -0,0 +1,238 @@
+limitations under the License.
+*/
+
+#ifndef MODULES_LLM_CACHE_DS_KV_CACHE_CHUNK_H_
+#define MODULES_LLM_CACHE_DS_KV_CACHE_CHUNK_H_
+
+#include <chrono>
+#include <condition_variable>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <regex>
+#include <shared_mutex>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "client/client.h"
+#include "client/ds/blob.h"
+#include "client/ds/i_object.h"
+#include "client/ds/remote_blob.h"
+#include "client/rpc_client.h"
+#include "llm-cache/ds/kv_tensor.h"
+
+namespace vineyard {
+
+// forward declaration
+struct LLMKV;
+
+// A KVCacheChunk contains all the KV tensors of a fixed number of
+// tokens (i.e., `chunk_size`).
+//
+// In its object blob, we first store all the KV tensors, and then
+// store all the tokens (including prefix tokens and current tokens
+// cached in the chunk), which will be used to avoid hash conflicts.
+//
+// In its metadata, we store the namespace (i.e., `ns_`), which will
+// be used as the name prefix of each chunk. Clients can also use the
+// namespace to list all the chunks. Access time (i.e., 'access_time_`)
+// in its metadata is used for the TTL-based global GC. We also have
+// the md5sum of all tokens (including prefix tokens and current tokens)
+// in its metadata. When we reconstruct a chunk from the object blob
+// and metadata, we calculate the md5sum of all tokens in the blob and
+// compare it with the md5sum in the metadata. If they are the same,
+// we consider the chunk is valid. Otherwise, we consider the chunk is
+// corrupted. By far, we don't use the md5sum of the tensors to alleviate
+// the compute overhead.
+class KVCacheChunk : public vineyard::Registered<KVCacheChunk> {
+ public:
+  inline static constexpr char kFieldNameNS[] = "namespace";
+  inline static constexpr char kFieldNameBuffer[] = "buffer";
+  inline static constexpr char kFieldNameTotalTokens[] = "total_tokens";
+  inline static constexpr char kFieldNameTensorNBytes[] = "tensor_nbytes";
+  inline static constexpr char kFieldNameLayer[] = "layer";
+  inline static constexpr char kFieldNameChunkSize[] = "chunk_size";
+  inline static constexpr char kFieldNameAccessTime[] = "access_time";
+  inline static constexpr char kFieldNameMd5[] = "md5";
+
+ private:
+  std::shared_ptr<Buffer> buffer_;
+  // number of prefix tokens and current tokens in the chunk
+  int total_tokens_;
+  int tensor_nbytes_;
+  int layer_;
+  int chunk_size_;
+  // access time is used for TTL-based global GC
+  uint64_t access_time_;
+  // md5sum of all tokens (including prefix tokens and current tokens)
+  std::string md5_;
+  // namespace. chunks within the same namespace will be shared
+  // among different clients
+  std::string ns_;
+
+ public:
+  static std::unique_ptr<Object> Create() __attribute__((used)) {
+    return std::static_pointer_cast<Object>(
+        std::unique_ptr<KVCacheChunk>{new KVCacheChunk()});
+  }
+
+  void Construct(const ObjectMeta& meta) override;
```

Comment:
IIUC, this method can only be called after `meta` has been populated by fetching object metadata & data (since you're populating buffers also)? If so, please clearly document this or refer readers to the base class for this requirement.